rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,223 @@
+and so are the parentheses around the functions.
+With the lower syntax overhead, using the operator becomes attractive in more situations,
+whereas a `pipe` pays for its overhead only in more complex scenarios (usually three functions or more).
+Having a dedicated operator also increases visibility and discoverability of the feature.
+
+# Detailed design
+[design]: #detailed-design
+
+## `|>` operator
+
+A new operator `|>` is introduced into the Nix language.
+Semantically, it is defined as the reverse of function application: `f a` = `a |> f`.
+It is left-associative and has a binding strength one weaker than function application:
+`a |> f |> g b |> h` = `h ((g b) (f a))`.
+
+## `builtins.pipe`
+
+`lib.pipe`'s functionality is implemented as a built-in function.
+The main motivation for this is that it allows to give better error messages
+like line numbers when some part of the pipeline fails.
+Additionally, it allows easy usage outside of Nixpkgs and increases discoverability.
+
+While Nixpkgs is bounds to minimum Nix versions and thus `|>` won't be available until
+several years after its initial implementation,
+it can directly benefit from `builtins.pipe` and its better error diagnostic by overriding `lib.pipe`.
+Elevating a Nixpkgs library function to a builtin has been done several times before,
+for example `bitAnd`, `splitVersion` and `concatStringsSep`.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+## Tooling support
+
+Like any language extension, this will require the available Nix tooling to be updated.
+Updating parsers should be pretty easy, as the syntax changes to the language are fairly minimal.
+Tooling that evaluates Nix code in some way or does static code analysis should be easy to support too,
+since one may treat the operator as syntactic sugar for function application.
+No fundamentally new semantics are introduced to the language.
+
+# Prior art
+
+Nickel has `|>` too, with the same name and semantics.
+
+F# has `|>`, called "pipe-forward" operator, with the same semantics.
+Additionally, it also has "pipe-backward" `<|` and `>>`/`<<` for forwards and backwards function composition.
+`<|` is equivalent to function application, however its lower binding order allows removing parentheses:
+`g (f a)` = `g <| f a`. All these operators have the same precedence and are left-associative.
+
+Elm has the same operators as F#.
+
+Haskell has the (backwards) function composition operator `.` in its prelude: `(g . f) a` = `g (f a)`.
+It also has "reverse application" `&`, which is roughly equivalent to `|>`,
+and `$`, which is function application again but right-associative and very weakly binding.
+`.` binds stronger than both.
+
+`|>` is definable as an infix function in several other programming languages,
+and in even more languages as macro or higher-order function (including Nix, that's `lib.pipe`).
+
+# Alternatives
```

Comment:
I'd like to propose an alternative that I already previously commented in https://github.com/NixOS/nix/issues/1845#issuecomment-1520798023 and https://github.com/NixOS/rfcs/pull/135#issuecomment-1557132711: To have an `apply` syntax like this:

```
apply f;
apply g;
x

# The same as
f (g x)
```

This is effectively the `$` operator (which is why I originally proposed it in https://github.com/NixOS/nix/issues/1845), but with more nixy syntax.

Advantages:
- More natural syntax, fitting better into the current Nix language, since there's already `with ...;` and `assert ...;`.
- This would be consistent with the new (experimental) CLI's [`--apply` flag](https://nixos.org/manual/nix/unstable/command-ref/new-cli/nix3-eval.html?highlight=--apply#options), which works in the same way.
- More verbose, making it clearer what the code does, and doesn't need a lot of explanation
- (weak) Doesn't add another operator, keeping the language simpler regarding operator precedence.

Disadvantages:
- More verbose, it takes longer to type
- The order is inverse from a pipe operator, making it harder to read in some cases.

Examples:
```
# Hypothetical `lib.guard`
apply lib.guard { ensure = x != 0; message = "Cannot divide by zero"; };
10 / x

apply builtins.trace "evaluating condition";
if condition then "foo" else "bar"

# Grouping and concatenating a list of strings by their 2-letter prefix:
apply mapAttrs (name: concatMapStringsSep " ");
apply groupBy (substring 0 2);
[ "foo" "bar" "baz" ]
```

Reply:
Adding to the disadvantages list:

- It would be yet another language feature that Nix reinvents in a custom way\*, instead of using constructs that users might recognize from other languages. (Or is there any prior work in other languages to this? It kind of reminds me of Rust function attributes and Python decorators a bit, but those are still significantly different)
- (big) Requires a new keyword in the language, which is a lot trickier w.r.t. backwards compatibility than introducing a new operator

I'll definitely add this to the alternatives list, but unless there is a way to solve the backwards compatibility issues without having to depend on the introduction of Nix language versioning, I remain skeptical.

\* To be fair some of these predate the equivalents seen elsewhere as it is a pretty old language, but for features introduced today "doing things differently" requires some pretty convincing arguments IMO.

Reply:
Yeah the backwards compatibility thing is a problem, but it's a more general problem that needs to be fixed at some point. https://github.com/NixOS/rfcs/pull/137 is being worked on by @fricklerhandwerk and @yorickvp (not committed though), which should pave a way towards being able to make such changes.

Reply:
> Or is there any prior work in other languages to this

`apply` tactic in theorem provers. But it's a rather ad-hoc kind of tactic, `refine` is more general and allows doing things like `refine func arg (λx => ?_)` which works like `use <- func(arg)` [use expression](https://gleam.run/news/v0.25-introducing-use-expressions/) in gleam.

Reply:
I think semicolon `;` already makes language much more confusing because the first thought that goes into it is that it's a statement end, rather than a separation like `,` (comma) in most other languages.

I would not recommend going anything with this semicolon approach, as this would better be attained by `let ...; in ...` anyway:

> ```
> apply f;
> apply g;
> x
> 
> # The same as
> f (g x)
> ```
@infinisil This doesn't make much sense other than breaking the function call into multiple lines, what we rather wanted to achieve with pipe operator was to invert the syntax so `input -> output` is clearly readable, rather than `f ( g <- input )` which has to be read as "input goes into g goes into f" to understand the flow.

Reply:
At least for me, this `apply` syntax is very confusing. My first thought when seeing `apply f` was, "The `apply` function is being called on `f`, but the result is not used anywhere. Does `apply` have side effects? That is strange because I though Nix was a purely functional language. I guess `apply` breaks the purity of Nix.". I would have no idea what was actually happening if your example did not include the equivalent in simple Nix syntax.

Reply:
Agree that it fits into the language syntax. 

But tbh. the syntax:

`foo = apply g; x;` 

seems unintuitive unless you know that `assert` requires two semicolons. 

I like the idea of having a pipe operator, over this syntax, because it feels more natural and follows established conventions in other languages (as mentioned in the rfc).

Reply:
> `assert` requires two semicolons.

Technically only one. The last one is for a binding or attribute, not for the `assert` expression.

`assert true; 1` is a valid expression, as you can see by entering it in the repl or into a file.

Kinda proves the point about `apply f; a` though.

Reply:
Shepherds currently believe this conversation can be resolved by adding a summary of this alternative to this section.